### ISSUE TITLE
[mlir][linalg-detensorize] Entry block arg types get converted.

### ISSUE THREAD

**Issue by silvasean** [llvm/llvm-project#62249]
Running:
```shell
mlir-opt --pass-pipeline='builtin.module(func.func(linalg-detensorize))' repro.mlir
```
Gives the following error:

```console
repro.mlir:3:3: error: 'func.func' op type of entry block argument #0('f32') must match the type of the corresponding argument in function signature('tensor<f32>')
```

Full error log:
https://gist.github.com/silvasean/73150502029639c261620bae4c8035f4

```mlir
#map = affine_map<() -> ()>
module {
  func.func @main(%arg0: tensor<f32>) -> tensor<f32> {
    %0 = tensor.empty() : tensor<f32>
    %1 = linalg.generic {indexing_maps = [#map, #map], iterator_types = []} ins(%arg0 : tensor<f32>) outs(%0 : tensor<f32>) {
    ^bb0(%in: f32, %out: f32):
      linalg.yield %in : f32
    } -> tensor<f32>
    cf.br ^bb1(%1 : tensor<f32>)
  ^bb1(%2: tensor<f32>):  // pred: ^bb0
    return %2 : tensor<f32>
  }
}
```

So in plain English the pass ends up changing the function entry block's arg type (from `tensor<f32>` to `f32`) is wrong because the pass [deliberately does not change function signatures](https://github.com/llvm/llvm-project/blob/a080798fb2d2343704512ad6adbf9fe1e13eec96/mlir/lib/Dialect/Linalg/Transforms/Detensorize.cpp#L496-L500), and specifically avoids modifying the entry block.

This seems to be some spooky action at a distance in the dialect conversion framework. Essentially what seems to happen is that:

1. [FunctionNonEntryBlockConversion](https://github.com/llvm/llvm-project/blob/a080798fb2d2343704512ad6adbf9fe1e13eec96/mlir/lib/Dialect/Linalg/Transforms/Detensorize.cpp#L97) converts the non-entry block, [registering](https://github.com/llvm/llvm-project/blob/a080798fb2d2343704512ad6adbf9fe1e13eec96/mlir/lib/Transforms/Utils/DialectConversion.cpp#L1328) the [DetensorizeTypeConverter](https://github.com/llvm/llvm-project/blob/a080798fb2d2343704512ad6adbf9fe1e13eec96/mlir/lib/Dialect/Linalg/Transforms/Detensorize.cpp#L140) as the type converter for the func.func's region
2. [DetensorizeGenericOp](https://github.com/llvm/llvm-project/blob/a080798fb2d2343704512ad6adbf9fe1e13eec96/mlir/lib/Dialect/Linalg/Transforms/Detensorize.cpp#L64) inlines the body of the linalg.generic, which triggers [OperationLegalizer::legalizePatternBlockActions](https://github.com/llvm/llvm-project/blob/a080798fb2d2343704512ad6adbf9fe1e13eec96/mlir/lib/Transforms/Utils/DialectConversion.cpp#L2005) which then [calls convertBlockSignature](https://github.com/llvm/llvm-project/blob/a080798fb2d2343704512ad6adbf9fe1e13eec96/mlir/lib/Transforms/Utils/DialectConversion.cpp#L2028) on the entry block (which is the block that the body was inlined into).

I'm really not sure what the right solution is here. One thing that is not clear to me is why the dialect conversion framework wants to change the block arg types of the destination block of inlining. That definitely seems very undesirable in this case.

**Comment by ftynse** [comment]
I suppose the type conversion is triggered so inlining a block doesn't create IR with mismatching types when the parent op of the region was doing some "casting" prior to inlining. One can argue that argument materialization should be performed instead, but this gets into the tricky interplay between various materializations and conversions.

Instead, I propose a simple workaround that temporarily creates a dummy entry block: https://reviews.llvm.org/D148983.
